### ISSUE TITLE
`hex2num` is vulnerable to overflow

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#22031]
From https://discourse.julialang.org/t/bigint-to-hex-and-back-to-bigint/3873.

The documentation simply says:

```
  hex2num(str)

  Convert a hexadecimal string to the floating point number it represents.
```

and does not hint on any limit of size in the allowed hex string.

`hex2num` should probably detect the overflow and try parsing the string as a `BigInt` before converting it?

**Comment by StefanKarpinski** [comment]
How about we delete this instead? It should be part of either parsing or reinterpret. In what sense does a hexadecimal string represent a floating-point number? cc @JeffBezanson; at the very least, the doc string should explain this.

**Comment by rfourquet** [comment]
I think the difference between `parse` and `hex2num` is the same as between `bin` and `bits` for integers, i.e. `hex2num` is concerned only by the machine bit-representation. Cf. #14418.

**Comment by StefanKarpinski** [comment]
We have `parse` to convert hex strings into 64-bit chunks of data and `reinterpret` to convert that into a `Float64`, so I'm not sure why this function is fundamental enough to exist in Base.

**Comment by simonbyrne** [comment]
+1 for getting rid of this: I think it's a Matlab/Octave-ism that has hung around. It's also not a great pattern to encourage, as it's technically endian-dependent (if you want lossless text-based serialisation, you should be using hex float literals).

**Comment by stevengj** [comment]
The package that seems to use this is [CBOR.jl](https://github.com/saurvs/CBOR.jl/blob/2658d364f8cfc912b62d39be0c3744c67504a721/src/decoding-common.jl#L187-L189), which somewhat perversely converts bytes to a string (via `bytes2hex`) and then to floating-point data, making at least 3 copies of the data in the process (array -> slice -> string -> float).

**Comment by simonbyrne** [comment]
It's also type-unstable.

**Comment by simonbyrne** [comment]
Also, `num2hex` doesn't appear to handle negative integers correctly.
```
julia> num2hex(-1)
"-0000000000000001"
```

**Comment by rfourquet** [comment]
> Also, num2hex doesn't appear to handle negative integers correctly.

This is fixed in #22039 (which addresses also the `hex2num` type-instability if the user is willing to specify a destination type).

**Comment by fredrikekre** [comment]
Given #22088, is this still relevant?
